### PR TITLE
br/lightning: add kv writer for external backend

### PR DESCRIPTION
**Pull request by tangenta**
### What problem does this PR solve?

Issue Number: ref #45719

Problem Summary:

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: br/pkg/lightning/backend/external/writer.go [hunk 1]
```diff
@@ -0,0 +1,331 @@
...
+	"github.com/pingcap/tidb/br/pkg/lightning/backend/encode"
+	"github.com/pingcap/tidb/br/pkg/lightning/backend/kv"
+	"github.com/pingcap/tidb/br/pkg/lightning/common"
+	"github.com/pingcap/tidb/br/pkg/membuf"
+	"github.com/pingcap/tidb/br/pkg/storage"
+	tidbkv "github.com/pingcap/tidb/kv"
+	"github.com/pingcap/tidb/util/logutil"
+	"github.com/pingcap/tidb/util/size"
+	"go.uber.org/zap"
+	"golang.org/x/exp/slices"
```
Comment (hawkingrei) [inline br/pkg/lightning/backend/external/writer.go:35]: You can use ```slices``` in the std package to replace ```exp/slices```.

File: br/pkg/lightning/backend/external/writer.go [hunk 2]
```diff
@@ -0,0 +1,331 @@
...
+	}
+
+	ts := time.Now()
+	var saveBytes uint64
+
+	defer func() {
+		w.currentSeq++
+		err := dataWriter.Close(w.ctx)
+		if err != nil {
+			logutil.BgLogger().Error("close data writer failed", zap.Error(err))
```
Comment (lance6716) [inline br/pkg/lightning/backend/external/writer.go:270]: this is a critical error, we should add some retry mechanism or return the error and wait it to be fixed in future PR

https://github.com/pingcap/tidb/issues/45353 happens when dumpling dump 100TB to s3

File: br/pkg/lightning/backend/external/writer.go [hunk 3]
```diff
@@ -0,0 +1,331 @@
...
+		}
+		err = statWriter.Close(w.ctx)
```
Comment (lance6716) [inline br/pkg/lightning/backend/external/writer.go:272]: the two closing is not atomic. what if TiDB node crashes between them and how will we find the dirty files when scanning s3? Do we treat the creating of stat file as the commit point of writing files to s3?

File: br/pkg/lightning/backend/external/writer.go [hunk 4]
```diff
@@ -0,0 +1,331 @@
...
+		if err != nil {
+			logutil.BgLogger().Error("close stat writer failed", zap.Error(err))
+		}
+		logutil.BgLogger().Info("flush kv",
```
Comment (lance6716) [inline br/pkg/lightning/backend/external/writer.go:276]: should use logger from ctx

File: br/pkg/lightning/backend/external/writer.go (hunk 3 above)
Comment (tangenta) [reply br/pkg/lightning/backend/external/writer.go:272]: Maybe we should add the validation check when TiDB reading from s3.

File: br/pkg/lightning/backend/external/writer.go [hunk 5]
```diff
@@ -0,0 +1,353 @@
...
+}
+
+type status bool
+
+// Flushed implements the backend.ChunkFlushStatus interface.
+func (s status) Flushed() bool {
+	return bool(s)
+}
+
+func (w *Writer) flushKVs(ctx context.Context) (err error) {
```
Comment (lance6716) [inline br/pkg/lightning/backend/external/writer.go:269]: what's the difference of this `ctx` and `w.ctx`?

File: br/pkg/lightning/backend/external/writer.go [hunk 6]
```diff
@@ -0,0 +1,353 @@
...
+	var kvSize uint64
+	for _, pair := range w.writeBatch {
+		err = w.kvStore.AddKeyValue(pair.Key, pair.Val)
+		if err != nil {
+			return err
+		}
+		kvSize += uint64(len(pair.Key)) + uint64(len(pair.Val))
+	}
+
+	if w.rc.currProp.keys > 0 {
```
Comment (lance6716) [inline br/pkg/lightning/backend/external/writer.go:322]: I prefer we always let `KeyValueStore` maintains the `rangeProperty`. We can add `Close()` to `KeyValueStore` to save the last `rangeProperty`

File: br/pkg/lightning/backend/external/writer.go [hunk 7]
```diff
@@ -0,0 +1,353 @@
...
+	if err != nil {
+		return err
+	}
+
+	w.recordMinMax(w.writeBatch[0].Key, w.writeBatch[len(w.writeBatch)-1].Key, kvSize)
+
+	w.writeBatch = w.writeBatch[:0]
+	w.rc.reset()
+	w.kvBuffer.Reset()
+	saveBytes = w.batchSize
```
Comment (lance6716) [inline br/pkg/lightning/backend/external/writer.go:336]:
```suggestion
	savedBytes = w.batchSize
```

File: br/pkg/lightning/backend/external/writer_test.go [hunk 8]
```diff
@@ -0,0 +1,92 @@
...
+	"github.com/pingcap/tidb/br/pkg/lightning/common"
+	"github.com/pingcap/tidb/br/pkg/storage"
+	"github.com/stretchr/testify/require"
+	"golang.org/x/exp/rand"
+	"golang.org/x/exp/slices"
+)
+
+func TestWriter(t *testing.T) {
+	seed := time.Now().Unix()
+	rand.Seed(uint64(1))
```
Comment (lance6716) [inline br/pkg/lightning/backend/external/writer_test.go:34]: you want to use seed?

**Review by lance6716 — COMMENTED** [review]
rest lgtm

File: br/pkg/lightning/backend/external/writer.go (hunk 3 above)
Comment (lance6716) [reply br/pkg/lightning/backend/external/writer.go]: Or maybe the framework will choose another `filenamePrefix` for the writer, and only expose the successful `filenamePrefix`s to following reader?

File: br/pkg/lightning/backend/external/writer.go [hunk 9]
```diff
@@ -0,0 +1,353 @@
...
+	onClose        OnCloseFunc
+
+	bufferPool *membuf.Pool
+}
+
+// NewWriterBuilder creates a WriterBuilder.
+func NewWriterBuilder() *WriterBuilder {
+	return &WriterBuilder{
+		memSizeLimit:   256 * size.MB,
+		writeBatchSize: 8 * 1024,
```
[the remark this answers is not in the document]
Comment (tangenta) [reply br/pkg/lightning/backend/external/writer.go:93]: This is the row count, I don't think it is appropriate to use `size.KB`. Let me change `writeBatchSize` to `writeBatchCount`

File: br/pkg/lightning/backend/external/writer.go [hunk 10]
```diff
@@ -0,0 +1,353 @@
...
+
+// SetBufferPool sets the buffer pool of the writer.
+func (b *WriterBuilder) SetBufferPool(bufferPool *membuf.Pool) *WriterBuilder {
+	b.bufferPool = bufferPool
+	return b
+}
+
+// Build builds a new Writer.
+func (b *WriterBuilder) Build(
+	ctx context.Context,
```
[the remark this answers is not in the document]
Comment (tangenta) [reply br/pkg/lightning/backend/external/writer.go:138]: Let me remove this inner context.

**Review by lance6716 — APPROVED** [review]

File: br/pkg/lightning/backend/external/file.go [hunk 11]
```diff
@@ -108,6 +108,14 @@ func (s *KeyValueStore) AddKeyValue(key, value []byte) error {
 	return nil
 }
 
+// Close closes the KeyValueStore and append the last range property.
+func (s *KeyValueStore) Close() {
```
Comment (lance6716) [inline br/pkg/lightning/backend/external/file.go:112]: modify an unit test in file_test.go to cover it.

Comment (lance6716) [reply br/pkg/lightning/backend/external/file.go:112]: I mean checking "append the last range property" really happens 😂 @tangenta

File: br/pkg/lightning/backend/external/writer.go [hunk 12]
```diff
@@ -0,0 +1,340 @@
...
+
+// IsSynced implements the backend.EngineWriter interface.
+func (w *Writer) IsSynced() bool {
+	return false
+}
+
+// Close closes the writer.
+func (w *Writer) Close(ctx context.Context) (backend.ChunkFlushStatus, error) {
+	if w.closed {
+		return status(true), nil
```
Comment (wjhuang2016) [inline br/pkg/lightning/backend/external/writer.go:216]: What's the situation w.closed is true?

File: br/pkg/lightning/backend/external/writer_test.go [hunk 13]
```diff
@@ -0,0 +1,92 @@
...
+	var keyCnt uint64 = 0
+	for {
+		p, err := statReader.nextProp()
+		if err == io.EOF {
+			break
+		}
+		require.NoError(t, err)
+		keyCnt += p.keys
+	}
+	require.Equal(t, uint64(kvCnt), keyCnt)
```
Comment (wjhuang2016) [inline br/pkg/lightning/backend/external/writer_test.go:91]: How about adding the test for two or three flush() and checking the data and stat file names?

File: br/pkg/lightning/backend/external/writer.go (hunk 12 above)
Comment (tangenta) [reply br/pkg/lightning/backend/external/writer.go:216]: Normally, it should be closed only once.

Comment (wjhuang2016) [reply br/pkg/lightning/backend/external/writer.go:216]: So, if it's a bug that w.closed is true, we can print an error log.

**Review by wjhuang2016 — APPROVED** [review]

**Pull request merged**